libbeat/common/kubernetes/util.go

```diff
@@ -92,60 +107,76 @@ func IsInCluster(kubeconfig string) bool {
 
 // DiscoverKubernetesNode figures out the Kubernetes node to use.
 // If host is provided in the config use it directly.
-// If beat is deployed in k8s cluster, use hostname of pod which is pod name to query pod meta for node name.
+// If it is empty then return discoverKubernetesNode.
+func DiscoverKubernetesNode(log *logp.Logger, configHost string, isInCluster bool, client kubernetes.Interface) (string, error) {
+	if configHost != "" {
+		log.Infof("kubernetes: Using node %s provided in the config", configHost)
+		return configHost, nil
+	}
+	hd := &hostDiscovery{}
+	d := &discoveryUtils{eDisc: hd, client: client, isInCluster: isInCluster}
+
+	return d.discoverKubernetesNode(log)
+}
+
+// discoverKubernetesNode figures out the Kubernetes node to use.
+// If beat is deployed in k8s cluster, use hostname of pod as the pod name to query pod metadata for node name.
 // If beat is deployed outside k8s cluster, use machine-id to match against k8s nodes for node name.
-func DiscoverKubernetesNode(log *logp.Logger, host string, inCluster bool, client kubernetes.Interface) (node string) {
-	if host != "" {
-		log.Infof("kubernetes: Using node %s provided in the config", host)
-		return host
+// If node cannot be discovered, return NODE_NAME env var as default value. In case it is not set return error.
+func (d *discoveryUtils) discoverKubernetesNode(log *logp.Logger) (string, error) {
```

Comment:
I think I'd write this method on reverse. I mean to pass in as arguments the stuff that is being used (like `eDisc HostDiscovery` and "compose" `discoveryUtils` with the logger so you don't have to pass down the stack the logger in every method call.

It's even possible that Beats shouldn't be using the logger at this level of the stack, but that will require a deeper analysis.